File: Day1-9/1.cpp
#include <fstream>
#include <iostream>
#include <vector>
using namespace std;

int fuelReq(int m)
{
    return (m / 3) - 2;
}

int part1(vector<int> &modules)
{
    int out = 0;
    for (auto m : modules)
    {
        out += fuelReq(m);
    }
    return out;
}

int part2(vector<int> &modules)
{
    int out = 0;
    for (auto m : modules)
    {
        int f = fuelReq(m);
        while (f > 0)
        {
            out += f;
            f = fuelReq(f);
        }
    }
    return out;
}

int main()
{
    vector<int> modules;

    ifstream infile("1.txt");
    string line;
    while (getline(infile, line))
    {
        modules.push_back(stoi(line));
    }

    cout << "Part 1: " << part1(modules) << "\n";
    cout << "Part 2: " << part2(modules) << "\n";

    return 0;
}

File: Day1-9/1.js
const fs = require("fs");

function fuelReq(m) {
  return Math.max(0, Math.floor(m / 3) - 2);
}

function part1(modules) {
  return modules.map(fuelReq).reduce((a, b) => a + b, 0);
}

function part2(modules) {
  let p2Fuel = m => {
    let f = fuelReq(m);
    if (f > 0) {
      f += p2Fuel(f);
    }
    return f;
  };

  return modules.map(p2Fuel).reduce((a, b) => a + b, 0);
}

function main() {
  let data = fs.readFileSync("1.txt");
  let modules = data
    .toString()
    .split("\n")
    .map(line => new Number(line));
  console.log("Part 1: " + part1(modules));
  console.log("Part 2: " + part2(modules));
}

main();
